File: concurrency/01116_print_zero_even_odd/custom_sem.c
void printNumber(int x);

struct custom_sem {
    uint8_t state;
    pthread_mutex_t mtx;
    pthread_cond_t cv;
};

struct custom_sem custom_sem_init(const uint8_t val)
{
    struct custom_sem init;
    init.state = val;
    pthread_mutex_init(&init.mtx, NULL);
    pthread_cond_init(&init.cv, NULL);
    return init;
}

static uint8_t custom_sem_wait(struct custom_sem* sem)
{
    pthread_mutex_lock(&sem->mtx);
    while (sem->state == 0) {
        pthread_cond_wait(&sem->cv, &sem->mtx);
    }
    const uint8_t res = sem->state;
    sem->state = 0;
    pthread_mutex_unlock(&sem->mtx);
    return res;
}

static void custom_sem_post(struct custom_sem* sem, const uint8_t state)
{
    pthread_mutex_lock(&sem->mtx);
    sem->state = state;
    pthread_cond_signal(&sem->cv);
    pthread_mutex_unlock(&sem->mtx);
}

typedef struct {
    int n;
    struct custom_sem states[3];
} ZeroEvenOdd;

ZeroEvenOdd* zeroEvenOddCreate(int n)
{
    ZeroEvenOdd* obj = (ZeroEvenOdd*) malloc(sizeof(ZeroEvenOdd));
    obj->n = n;
    obj->states[0] = custom_sem_init(1);
    obj->states[1] = custom_sem_init(0);
    obj->states[2] = custom_sem_init(0);
    return obj;
}

// You may call global function `void printNumber(int x)`
// to output "x", where x is an integer.

void zero(ZeroEvenOdd* obj)
{
    const int n = obj->n;
    for (int i = 0; i < n; i += 1) {
        const uint8_t child = custom_sem_wait(&obj->states[0]);
        printNumber(0);
        custom_sem_post(&obj->states[child], 1);
    }
}

void odd(ZeroEvenOdd* obj)
{
    const int n = obj->n;
    for (int i = 1; i <= n; i += 2) {
        custom_sem_wait(&obj->states[1]);
        printNumber(i);
        custom_sem_post(&obj->states[0], 2);
    }
}

void even(ZeroEvenOdd* obj)
{
    const int n = obj->n;
    for (int i = 2; i <= n; i += 2) {
        custom_sem_wait(&obj->states[2]);
        printNumber(i);
        custom_sem_post(&obj->states[0], 1);
    }
}

void zeroEvenOddFree(ZeroEvenOdd* obj)
{
    pthread_mutex_destroy(&obj->states[0].mtx);
    pthread_mutex_destroy(&obj->states[1].mtx);
    pthread_mutex_destroy(&obj->states[2].mtx);
    pthread_cond_destroy(&obj->states[0].cv);
    pthread_cond_destroy(&obj->states[1].cv);
    pthread_cond_destroy(&obj->states[2].cv);
    free(obj);
}

File: concurrency/01116_print_zero_even_odd/custom_sem.rs
use std::sync::Condvar;

struct Sem {
    mtx: Mutex<u8>,
    cv: Condvar,
}

impl Sem {
    fn new(init_state: u8) -> Self {
        Self {
            mtx: Mutex::new(init_state),
            cv: Condvar::new(),
        }
    }

    fn wait(&self) -> u8 {
        let mut count = self.mtx.lock().unwrap(); 
        while *count == 0 {
            count = self.cv.wait(count).unwrap();
        }
        let ret_val = *count;
        *count = 0;
        ret_val
    }

    fn post(&self, state: u8) {
        let mut count = self.mtx.lock().unwrap();
        *count = state;
        self.cv.notify_one();
    }
}

struct ZeroEvenOdd {
    n: i32,
    states: [Sem; 3],
}

impl ZeroEvenOdd {
    fn new(n: i32) -> Self {
        assert!(n >= 1);
        ZeroEvenOdd {
            n: n,
            states: [Sem::new(1), Sem::new(0), Sem::new(0)],
        }
    }

    // print_number(x) prints the integer x
    fn zero<F>(&self, print_number: F)
    where
        F: Fn(i32),
    {
        for _ in 0..self.n {
            let state = self.states[0].wait();
            print_number(0);
            self.states[state as usize].post(1);
        }
    }

    fn odd<F>(&self, print_number: F)
    where
        F: Fn(i32),
    {
        for curr in (1..=self.n).step_by(2) {
            self.states[1].wait();
            print_number(curr);
            self.states[0].post(2);
        }
    }

    fn even<F>(&self, print_number: F)
    where
        F: Fn(i32),
    {
        for curr in (2..=self.n).step_by(2) {
            self.states[2].wait();
            print_number(curr);
            self.states[0].post(1);
        }
    }
}
